src/workspace/mod.rs

```diff
@@ -0,0 +1,1823 @@
+        let root_dir = new_rc(Url::from_directory_path(root_dir).unwrap());
+        let mut config_folders = members
+          .into_iter()
+          .map(|(folder_url, config)| {
+            (
+              folder_url,
+              match config {
+                DenoOrPkgJson::Deno(config_file) => FolderConfigs {
+                  deno_json: Some(config_file),
+                  pkg_json: None,
+                },
+                DenoOrPkgJson::PkgJson(pkg_json) => FolderConfigs {
+                  deno_json: None,
+                  pkg_json: Some(pkg_json),
+                },
+              },
+            )
+          })
+          .collect::<IndexMap<_, _>>();
+        config_folders
+          .entry(root_dir.clone())
+          .or_default()
+          .deno_json = Some(root);
+        Workspace {
+          root_dir,
+          start_dir,
+          config_folders,
+        }
+      }
+    };
+    match maybe_npm_discovery {
+      PackageJsonDiscovery::Single(pkg_json) => {
+        let pkg_json_dir = new_rc(
+          Url::from_directory_path(pkg_json.path.parent().unwrap()).unwrap(),
+        );
+        if workspace
+          .root_dir
+          .as_str()
+          .starts_with(pkg_json_dir.as_str())
+        {
+          // the package.json was in a higher up directory
+          workspace.root_dir = pkg_json_dir.clone();
+        }
+
+        workspace
+          .config_folders
+          .entry(pkg_json_dir)
+          .or_default()
+          .pkg_json = Some(pkg_json);
+      }
+      PackageJsonDiscovery::Workspace { root, members } => {
+        let pkg_json_dir = new_rc(
+          Url::from_directory_path(root.path.parent().unwrap()).unwrap(),
+        );
+        if workspace
+          .root_dir
+          .as_str()
+          .starts_with(pkg_json_dir.as_str())
+        {
+          // the package.json was in a higher up directory
+          workspace.root_dir = pkg_json_dir.clone();
+        }
+
+        workspace
+          .config_folders
+          .entry(pkg_json_dir)
+          .or_default()
+          .pkg_json = Some(root);
+
+        for (member_dir, pkg_json) in members {
+          workspace
+            .config_folders
+            .entry(member_dir)
+            .or_default()
+            .pkg_json = Some(pkg_json);
+        }
+      }
+      PackageJsonDiscovery::None => {
+        // do nothing
+      }
+    }
+
+    debug_assert!(
+      workspace.config_folders.contains_key(&workspace.root_dir),
+      "root should always have a folder"
+    );
+    Ok(workspace)
+  }
+
+  pub async fn create_resolver<
+    TReturn: Future<Output = Result<String, AnyError>>,
+  >(
+    &self,
+    options: CreateResolverOptions,
+    fetch_text: impl Fn(&Url) -> TReturn,
+  ) -> Result<WorkspaceResolver, WorkspaceResolverCreateError> {
+    WorkspaceResolver::from_workspace(self, options, fetch_text).await
+  }
+
+  pub fn diagnostics(&self) -> Vec<WorkspaceDiagnostic> {
```

Comment:
Off-topic to this PR: should we create `deno_diagnostics` crate that would provide a common interface for all our crates that provide diagnostics?

Reply:
We do in deno_ast. I'm not sure it's worth using that here because that's for more complex scenarios. These are just simple text warnings that the CLI will log out.